### PR TITLE
Add support for docx, pptx and xlsx file types

### PR DESCRIPTION
**Pull request by forivall**
A re-implementation of #57, using the libmagic reference as the guide. Causes some linter warnings with the deep nesting. I used the same approach as the mkv detection, and, following the libmagic definition, won't read more than 3008 bytes.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -98,27 +98,72 @@ module.exports = input => {
...
+			// toHexArray('[Content_Types].xml')
+			// toHexArray('_rels/.rels')
+
+			// https://github.com/file/file/blob/master/magic/Magdir/msooxml
+			const sliced = buf.subarray(4, 4 + 2000);
+			const header2Pos = sliced.findIndex((el, i, arr) => arr[i] === 0x50 && arr[i + 1] === 0x4B && arr[i + 2] === 0x3 && arr[i + 3] === 0x4);
+
+			if (header2Pos >= 0) {
+				const slicedAgain = buf.subarray(header2Pos + 8, header2Pos + 8 + 1000);
+				const header3Pos = slicedAgain.findIndex((el, i, arr) => arr[i] === 0x50 && arr[i + 1] === 0x4B && arr[i + 2] === 0x3 && arr[i + 3] === 0x4);
```
Comment (kevva) [inline index.js]: This is essentially the same as `header2Pos`. Can you extract it to a function that takes an array as input instead?

File: index.js [hunk 2]
```diff
@@ -98,27 +98,72 @@ module.exports = input => {
...
+				ext: 'xpi',
+				mime: 'application/x-xpinstall'
+			};
+		}
+
+		if (
+			check([0x5B, 0x43, 0x6F, 0x6E, 0x74, 0x65, 0x6E, 0x74, 0x5F, 0x54, 0x79, 0x70, 0x65, 0x73, 0x5D, 0x2E, 0x78, 0x6D, 0x6C], {offset: 30}) ||
+			check([0x5F, 0x72, 0x65, 0x6C, 0x73, 0x2F, 0x2E, 0x72, 0x65, 0x6C, 0x73], {offset: 30})
+		) {
+			// Generating the above hex arrays:
```
Comment (kevva) [inline index.js]: Can you extract the following logic to a function? Having it inline makes it hard to read.

**Comment by kevva** [comment]
Really nice work :+1:

File: index.js [hunk 3]
```diff
@@ -98,27 +98,72 @@ module.exports = input => {
...
+			// toHexArray = s => '[' + Array.from(s).map((c) => '0x' + c.charCodeAt(0).toString(16).toUpperCase()).join(', ') + ']'
+			// toHexArray('[Content_Types].xml')
+			// toHexArray('_rels/.rels')
+
+			// https://github.com/file/file/blob/master/magic/Magdir/msooxml
+			const sliced = buf.subarray(4, 4 + 2000);
+			const header2Pos = sliced.findIndex((el, i, arr) => arr[i] === 0x50 && arr[i + 1] === 0x4B && arr[i + 2] === 0x3 && arr[i + 3] === 0x4);
```
Comment (kevva) [inline index.js]: Isn't `arr[i]` the same as `el`?

File: index.js (hunk 2 above)
Comment (forivall) [inline index.js]: Do you mean the commented `toHexArray` function or the `sliced.findIndex` stuff?

File: test.js [hunk 4]
```diff
@@ -72,7 +72,10 @@ const types = [
 	'mxf',
 	'mts',
 	'blend',
-	'bpg'
+	'bpg',
+	'docx',
+	'pptx',
+	'xlsx'
```
Comment (sindresorhus) [inline test.js]: Can you add these to the readme as links and to the package.json as keywords?

File: index.js [hunk 5]
```diff
@@ -98,27 +98,72 @@ module.exports = input => {
...
+				mime: 'application/x-xpinstall'
+			};
+		}
+
+		if (
+			check([0x5B, 0x43, 0x6F, 0x6E, 0x74, 0x65, 0x6E, 0x74, 0x5F, 0x54, 0x79, 0x70, 0x65, 0x73, 0x5D, 0x2E, 0x78, 0x6D, 0x6C], {offset: 30}) ||
+			check([0x5F, 0x72, 0x65, 0x6C, 0x73, 0x2F, 0x2E, 0x72, 0x65, 0x6C, 0x73], {offset: 30})
+		) {
+			// Generating the above hex arrays:
+			// toHexArray = s => '[' + Array.from(s).map((c) => '0x' + c.charCodeAt(0).toString(16).toUpperCase()).join(', ') + ']'
```
Comment (sindresorhus) [inline index.js]: Can you change to generate it live? We just do it once at module loading, so it won't have any performance impact.

File: index.js (hunk 2 above)
Comment (kevva) [inline index.js]: Yeah, the `slice` stuff.

**Comment by arhajiismael206** [comment]
I'm also interested in being able to identify `.xlsx` files. I can help test once this is ready.

File: test.js (hunk 4 above)
Comment (forivall) [inline test.js]: [Done!!](https://render.bitstrips.com/v2/cpanel/10120168-302682443_5-s1-v1.png?transparent=1&palette=1&width=246)

Comment (sindresorhus) [inline test.js]: @forivall I don't see the readme changes.

File: index.js [hunk 6]
```diff
@@ -98,27 +103,64 @@ module.exports = input => {
...
+			};
+		}
+
+		// https://github.com/file/file/blob/master/magic/Magdir/msooxml
+		if (check(oxmlContentTypes, {offset: 30}) || check(oxmlRels, {offset: 30})) {
+			const sliced = buf.subarray(4, 4 + 2000);
+			const nextZipHeaderIndex = arr => arr.findIndex((el, i, arr) => arr[i] === 0x50 && arr[i + 1] === 0x4B && arr[i + 2] === 0x3 && arr[i + 3] === 0x4);
+			const header2Pos = nextZipHeaderIndex(sliced);
+
+			if (header2Pos >= 0) {
```
Comment (kevva) [inline index.js]: Nitpick, but this is better written as `!== 0` imo.

Comment (forivall) [inline index.js]: this would be `!== -1` though. i can still make that change though

File: test.js (hunk 4 above)
Comment (forivall) [inline test.js]: whoops, was in a rush.

**Comment by forivall** [comment]
Readme updated, and `findIndex` checks now use `!== -1`

**Comment by kevva** [comment]
This looks good to go :+1:

**Comment by sindresorhus** [comment]
Thank you @forivall. This is great 🙌

**Pull request merged**